crates/autopilot/src/domain/fee/mod.rs

```diff
@@ -92,3 +74,89 @@ pub enum Policy {
         factor: f64,
     },
 }
+
+impl Policy {
+    pub fn with(
+        self,
+        order: &boundary::Order,
+        quote: &domain::Quote,
+        skip_market_orders: bool,
+    ) -> Option<Policy> {
+        match order.metadata.class {
+            boundary::OrderClass::Market => {
+                if skip_market_orders {
+                    None
+                } else {
+                    Some(self)
+                }
+            }
+            boundary::OrderClass::Liquidity => None,
+            boundary::OrderClass::Limit => {
+                if !skip_market_orders {
+                    Some(self)
+                } else {
+                    let order_ = boundary::Amounts {
+                        sell: order.data.sell_amount,
+                        buy: order.data.buy_amount,
+                        fee: order.data.fee_amount,
+                    };
+                    let quote_ = boundary::Amounts {
+                        sell: quote.sell_amount,
+                        buy: quote.buy_amount,
+                        fee: quote.fee,
+                    };
+                    if boundary::is_order_outside_market_price(&order_, &quote_) {
+                        Some(self)
+                    } else {
+                        None
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum PolicyBuilder {
```

Comment:
I think we should make this class private and create it in the main Fee component from the infra config. It might also be nicer to have every variant be there own sub-struct so that we can implement slightly different build behaviors (e.g. surplus and volume don't need the quote).

Then the main `apply` function would become something like

```rs
self.fee_variants.filter_map(|variant| {
  match variant {
    FeeVariant::Surplus(variant) => variant.apply(order),
    FeeVariant::PriceImprovement(variant) => variant.apply(order, quote),
    ...
  }
}
```

Reply:
I suppose, `variant.apply` function should return `Option<Policy>` and also should contain [this](https://github.com/cowprotocol/services/blob/c519fb8223a277291364ee4ec457aed128ac88e1/crates/autopilot/src/domain/fee/mod.rs#L79-L116) logic inside, but this requires `quote` regardless the Policy kind.

Reply:
Refactored